cpp/include/raft/core/detail/copy.hpp

```diff
@@ -0,0 +1,510 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+#include <cstdio>
+#include <execution>
+#include <raft/core/cuda_support.hpp>
+#include <raft/core/device_mdspan.hpp>
+#include <raft/core/error.hpp>
+#include <raft/core/host_mdspan.hpp>
+#include <raft/core/logger.hpp>
+#include <raft/core/mdspan.hpp>
+#include <raft/core/resource/stream_view.hpp>
+#include <raft/core/resources.hpp>
+#include <type_traits>
+#ifndef RAFT_DISABLE_CUDA
+#include <raft/core/cudart_utils.hpp>
+#include <raft/core/device_mdarray.hpp>
+#include <raft/core/resource/cublas_handle.hpp>
+#include <raft/linalg/detail/cublas_wrappers.hpp>
+#ifdef __CUDACC__
+#include <raft/util/cuda_dev_essentials.cuh>
+#endif
+#endif
+
+namespace raft {
+namespace detail {
+
+template <bool B, typename DstType = void, typename SrcType = void, typename T = void>
+struct mdspan_copyable {};
+
+/*
+ * A helper struct used to determine whether one mdspan type can be copied to
+ * another and if so how
+ */
+template <typename DstType, typename SrcType, typename T>
+struct mdspan_copyable<true, DstType, SrcType, T> {
```

Comment:
Here's how I would implement optimized mdspan copy.

## Dispatch logic

1. If the destination has a nonunique layout, then forbid copying (except for specializations for known nonunique layouts).

2. If the two mdspans have the same layout, then

    a. if the two mdspans have contiguous layout, then dispatch to same-contiguous-copy;

    b. otherwise, dispatch to same-noncontiguous-copy.

3. If the two mdspans have different layout, then

    a. if the layouts and accessors work with `cudaMemcpy*Async`, then dispatch to different-layout-known-copy;

    b. otherwise, dispatch to generic-flattened-copy.

## same-contiguous-copy

1. If it's valid to use `memcpy` (or `cudaMemcpy`, etc.) for the two accessors, then use it.

2. Otherwise, dispatch to same-noncontiguous-copy.

## same-noncontiguous-copy

1. If it's valid to use a parallel algorithm (or device kernel) with the data, then use generic flat copy from `[0, src.mapping().required_span_size())` (transformed through `src.accessor().access(src.data_handle(), _)`) to `[0, dst.mapping().required_span_size())` (transformed analogously through `dst.accessor().access(dst.data_handle(), _)`).

2. Otherwise, use `std::copy` with the same ranges.

## different-layout-known-copy

Use `cudaMemcpy*Async` (where `*` is `2D` or `3D`).

## generic-flattened-copy

Create a `layout_left::mapping` with the same extents as `src` and `dst`.  Invert it using divmods.  The composition of this inverse with the layout mapping of `src` gives a flat (1-D) range for `src`.  Ditto for `dst`.  Apply the same technique as in same-noncontiguous-copy.

Reply:
> If the destination has a nonunique layout, then forbid copying
Yep, I think that makes sense. That's a fine check to add to the `mdspan_copyable` helper.

> If it's valid to use memcpy (or cudaMemcpy, etc.) for the two accessors, then use it.
I believe this is exactly what we do already.

> If it's valid to use a parallel algorithm (or device kernel) with the data, then use generic flat copy from [0, src.mapping().required_span_size()) (transformed through src.accessor().access(src.data_handle(), _)) to [0, dst.mapping().required_span_size()) (transformed analogously through dst.accessor().access(dst.data_handle(), _)).

Yep, I thought about this approach, and I don't mind it, but it increases complexity a bit and will _sometimes_ decrease performance. Let me try it out with the benchmarks and just see what domains it actually provides benefit for and how much.

> Use cudaMemcpy*Async (where * is 2D or 3D).
I'm probably missing something, but I don't see how to do transposes with this.

> Create a layout_left::mapping with the same extents as src and dst. Invert it using divmods. The composition of this inverse with the layout mapping of src gives a flat (1-D) range for src. Ditto for dst. Apply the same technique as in same-noncontiguous-copy.

This is not guaranteed to iterate on elements in memory-order, and the current implementation is designed to limit the use of divmods for device code for performance reasons.

Reply:
> This is not guaranteed to iterate on elements in memory-order, and the current implementation is designed to limit the use of divmods for device code for performance reasons.

Right -- this is not the ideal case.  It's a reasonable fall-back for something like `linalg::copy` in P1673.  RAFT controls the resource type, so it has the option to maintain storage for packing and unpacking.

CUTLASS has code that automatically decomposes copies into minimal sequences of hardware-specific copy instructions, if it knows enough about the layouts at compile time.  One could imagine adapting that to work better with run-time information, but it would be a lot of work.

Reply:
> Right -- this is not the ideal case. It's a reasonable fall-back

Yep, makes sense. That's why the current kernel does not try to guarantee maximal coalescing; it's just a best effort setup that will ensure coalescing on both sides (read/write) in the most common cases and on at least one side on the other.